File: src/test/java/pages/purchasePages/PurchaseOrderPage.java
package pages.purchasePages;

import com.microsoft.playwright.FrameLocator;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.LoadState;
import com.microsoft.playwright.options.WaitForSelectorState;
import enums.Categories;
import enums.DistributorInfo;
import org.junit.Assert;
import pages.commonPages.BasePage;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.concurrent.ThreadLocalRandom;

public class PurchaseOrderPage extends BasePage {

    Locator pageTitle = page.locator("#PageTitle");
    Locator calendar = page.locator(".k-icon.k-i-calendar");
    Locator selectToday = page.locator(".k-link.k-nav-today");
    Locator clickDropdownToggle = page.locator("span.k-select > span.k-icon.k-i-arrow-s");
    FrameLocator frameLocator = page.frameLocator("iframe.k-content-frame");
    FrameLocator orderProductIdentificationFrame = getFrameByDialogTitle("Sipariş Ürünü Tanımlama");
    FrameLocator productDescriptionFrame = getFrameByDialogTitle("Ürün Tanımlama");

    /**
     * Sipariş Oluşturma sayfasını verify eder.
     */
    public void verifyCreateOrderPage() {
        verifyTextElementUseTrim("Sipariş İşlemleri", pageTitle);
    }

    /**
     * Sipariş Oluşturma sayfasında Sipariş Tarihi doldurur.
     */
    public void fillOrderCreationDate() {
        clickElement(calendar.nth(0));
        selectToday.click(new Locator.ClickOptions().setForce(true));
        //kendo component nedeni ile SimpleDateFormat işe yaramıyor!!!
    }

    /**
     * Sipariş Oluşturma sayfasında Sipariş Adı doldurur.
     */
    public void fillOrderNameOrderCreationPage() {
        String timestamp = new SimpleDateFormat("yyyyMMddHHmm").format(new Date()); // Zaman bazlı sayaç
        String orderName = "KMRN_TST_AUTO_" + timestamp;
        page.locator("#PurchaseOrderName").fill(orderName);
    }

    /**
     * Kategori listesini açar.
     */
    public void openCategoryList(){
        clickElement(clickDropdownToggle.nth(0));
    }

    /**
     * Sipariş Oluşturma sayfasında Kategori alanını seçer.
     */
    public void selectCategoryFromList(String category) {
        Locator selectCategory = page.locator("#CategoryId_listbox li[role='option'].k-item",
                new Page.LocatorOptions().setHasText(category));
        selectCategory.click(new Locator.ClickOptions().setForce(true));
    }
    /**
     * Kategori alanını doğrular.
     */
    public void verifyCategory(String category) {
        Locator selectedCategory = page.locator(".k-dropdown .k-input").nth(0);
        verifyTextElementUseTrim(category, selectedCategory);
    }

    /**
     * Sipariş Oluşturma sayfasında Dağıtıcı Firma frame'ini açar.
     */
    public void openCompanyIdentificationFrame() {
        clickElement(page.locator("#FirmIdButtonId"));
    }

    /**
     * Firma Tanımlama frame'inde Firma Kodu alanını doldurur.
     */
    public void fillCompanyCode(String category){
        Categories categoryLabel = Categories.fromLabel(category);
        if (categoryLabel != null){
            DistributorInfo distributorInfo = categoryLabel.getDistributorInfo();
            frameLocator.locator("#FilterFirmCode").fill(distributorInfo.getFirmCode());
        }
    }

    /**
     * Firma Tanımlama frame sorgula butonuna tıklar.
     */
    public void clickFilterButtonId(){
        frameLocator.locator("#FilterButtonId").click();
    }

    /**
     * Firma Tanımlama frame'inde sorgu sonucunda gelen firmayı seçer.
     */
    public void selectDistributorCompany(){
        frameLocator.locator("input[name^='FirmGridId']").nth(0).click();
    }

    /**
     * Sipariş Oluşturma sayfasında Firma İlgili Kişi  tıklar.
     */
    public void clickCompanyContactPerson(){
        clickElement(page.locator(".k-multiselect-wrap.k-floatwrap").nth(3));
    }

    /**
     * Sipariş Oluşturma sayfasında Firma İlgili Kişi  seçer.
     */
    public void selectCompanyContactPerson() {
        page.waitForSelector("#FirmResponsibleUserId_listbox");
        clickElement(page.locator("#FirmResponsibleUserId_option_selected").nth(0));
    }

    /**
     * Sipariş Oluşturma sayfasında Dağılım Hedef Tipi açar.
     */
    public void openDistributionTargetType(){

        clickDropdownToggle.nth(1).click(new Locator.ClickOptions().setForce(true));
    }

    /**
     * Sipariş Oluşturma sayfasında Dağılım Hedef Tipi seçer.
     */
    public void selectDistributionTargetType() {

        page.waitForSelector("#DistributionTargetTypeId_listbox li");
        clickElement(page.locator("#DistributionTargetTypeId_listbox li").nth(1));
    }

    /**
     * Sipariş Oluşturma sayfasında Giriş Antrepo frame'ini açar.
     */
    public void openWarehouseDefinitionFrame() {
        clickElement(page.locator("#EntryWarehouseIdButtonId"));
    }

    /**
     * Antrepo Tanımlama frame'inde Setur Bölgesi listesini açar.
     */
    public void openSeturRegionFields(){
        frameLocator.locator("span.k-select > span.k-icon.k-i-arrow-s")
                .nth(0).click();
    }

    /**
     * Antrepo Tanımlama frame Setur Bölgesi seçer.
     * @param region setur bölgesi
     */
    public void selectSeturRegionFromList(String region) {
        Locator allItems = frameLocator.locator("ul#FilterSeturRegionID_listbox li");
        Locator targetItem = allItems.filter(new Locator.FilterOptions().setHasText(region));
        targetItem.click(new Locator.ClickOptions().setForce(true));
    }
    /**
     * Antrepo Tanımlama frame Setur Bölgesi doğrular.
     * @param region setur bölgesi
     */
    public void verifySeturRegion(String region) {
        Locator selectedRegion = frameLocator.locator(".k-dropdown .k-input").nth(0);
        verifyTextElementUseTrim(region, selectedRegion);
    }

    /**
     * Antrepo Tanımlama frame'inde sorgu sonucunda gelen antrepoyu seçer.
     */
    public void selectWarehouse(){
        frameLocator.locator("input[name^='WarehouseGridId']").nth(0).click();
    }

    /**
     * Sipariş Oluşturma sayfasında Fatura Adresini açar.
     */
    public void openInvoiceAddress(){
        clickElement(clickDropdownToggle.nth(3));
    }

    /**
     * Sipariş Oluşturma sayfasında Fatura Adresini seçer.
     */
    public void selectInvoiceAddress() {
        page.waitForSelector("#CompanyAddressId_listbox li");
        clickElement( page.locator("#CompanyAddressId_listbox li").nth(1));
    }

    /**
     * Sipariş Oluşturma sayfasında Teslimat Adresini açar.
     */
    public void openDeliveryAddress(){
        clickElement(clickDropdownToggle.nth(4));

    }

    /**
     * Sipariş Oluşturma sayfasında Teslimat Adresini seçer.
     */
    public void selectDeliveryAddress() {
        page.waitForSelector("#WarehouseAddressId_listbox li");
        clickElement(page.locator("#WarehouseAddressId_listbox li").nth(1));
    }

    /**
     * Sipariş Oluşturma sayfasında Sipariş Otomatik Olarak Tamamlansın mı? Hayır seçer.
     */
    public void checkOrderCompleteAutomatically() {
        clickElement( page.locator("#no_CanAutoComplete"));
    }

    /**
     * Sipariş Oluşturma sayfasında siparişi kaydeder.
     */
    public void saveOrder(){
        clickElement(page.locator("#SaveBtn"));
    }
    /**
     * Kaydedilen siparişi doğrular.
     */
    public void verifyPurchaseOrderTabs() {
        page.waitForSelector("#PurchaseOrderTabs");
        verifyIsVisible(page.locator("#PurchaseOrderTabs"));
    }

    /**
     * Sipariş Numarasını doğrular.
     */
    public void verifyOrderByOrderCode(){
        page.waitForLoadState(LoadState.DOMCONTENTLOADED);
        String orderID = page.locator("#PurchaseOrderCode").getAttribute("value");
        Assert.assertNotNull("Order ID null olmamalı!", orderID);
        addString("orderCode", orderID);
    }

    /**
     * Ürün eklemek için Yeni Kayıt butonuna tıklar ve Sipariş Ürünü Tanımlama frame'ini açar.
     */
    public void openOrderProductDescriptionFrame() {
        clickElement(page.locator("a.k-grid-PurchaseOrderProductGridIdAddNew"));
    }

    /**
     * Ürün eklemek için Ürün Tanımlama frame'ini açar.
     */
    public void openProductDescriptionFrame(){
        //orderProductIdentificationFrame.locator("#ProductIdButtonId").click();
        frameLocator.locator("#ProductIdButtonId").click();
    }

    /**
     * Ürün Tanımlama frame'inde Ürün Adını alır ve kaydeder.
     */
    public void getProductName() {
        Locator productNameLocator = productDescriptionFrame.
                locator("td[data-field-name='ProductNameLong']").nth(0);

        productNameLocator.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE));
        String productName = productNameLocator.textContent().trim();
        System.out.println("Ürün adı: " + productName);
        addString("productName", productName);
    }

    /**
     * Ürün Tanımlama frame'inde sorgulama sonucunda gelen ürünü seçer.
     */
    public void selectProduct(){
        //frame içinde frame (nested iframe) bu nedenle frameLocator kullanılamaz
        productDescriptionFrame.locator("input[name^='ProductGrid']").nth(0).click();
    }

    /**
     * Sipariş Ürünü Tanımlama frame Satın Alma Para Birimi değerini alır.
     */
    public void getCurrencyCodes() {
        Locator currencyLocator = orderProductIdentificationFrame.locator("#ProductCurrencyCode");
        currencyLocator.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE));

        String productCurrency = (String) currencyLocator
                .evaluate("el => el.value || el.placeholder || el.textContent || el.innerText");

        addString("productCurrency", productCurrency);
    }


    /**
     * Sipariş Ürünü Tanımlama Sipariş- Ürün para birimleri uyuşmuyorsa frame kapatır.
     */
    public boolean ifCurrencyNotMatchCloseFrame() {
        String currencyCode = orderProductIdentificationFrame.locator("#CurrencyCode").inputValue();
        String productCurrency = getString("productCurrency");

        if (!currencyCode.equals(productCurrency)) {
            page.locator(".k-window-actions .k-window-action").nth(0).click();
            return true; // eşleşmiyor
        }
        return false; // eşleşiyor
    }

    /**
     * Sipariş Para Birimi listesini açar.
     */
    public void openOrderCurrencyCodes() {
        Locator currencyDropdownToggle = page.locator("span.k-select > span.k-icon.k-i-arrow-s");

        currencyDropdownToggle.nth(7).hover();
        currencyDropdownToggle.nth(7).click();
    }

    /**
     * Ürüne uyumlu olan para birimini seçer.
     */
    public void selectCurrencyCode() {
        String productCurrency = getString("productCurrency");

        Locator currencyOption = page.locator("ul#CurrencyCode_listbox li",
                new Page.LocatorOptions().setHasText(productCurrency));
        currencyOption.click(new Locator.ClickOptions().setTimeout(3000).setForce(true));

        //verifyTextElementUseTrim(productCurrency, currencyOption);
    }
    /**
     * Sipariş Ürünü Tanımlama frame'inde seçilen para birimini doğrular.
     */
    public void verifyProductCurrencyCode() {
        Locator selectedCurrency = page.locator(".k-dropdown .k-input").nth(3);
        String productCurrency = getString("productCurrency");
        verifyTextElementUseTrim(productCurrency, selectedCurrency);
    }


    public void confirmPopup() {
        page.locator(".ajs-button.ajs-ok").click();
    }

    /**
     * Sipariş Ürünü Tanımlama frame'inde Sipariş Adedi girer.
     */
    public void enterQuantityForProduct(){
        int randomQuantity = ThreadLocalRandom.current().nextInt(1, 21);
        setKendoNumericTextBoxValue(orderProductIdentificationFrame,"#Quantity",
                String.valueOf(randomQuantity));
        orderProductIdentificationFrame.locator("#Quantity").press("Enter");
    }

    /**
     * Sipariş Ürünü Tanımlama frame'inde Kaydet butonuna tıklar.
     */
    public void saveOrderProductsDescription(){
        orderProductIdentificationFrame.locator("#SaveBtn").click();
    }

    /**
     * Siparişe eklenen ürünü verify eder.
     */
    public void verifyProducts() {
        Locator tableProduct = page.locator("td[data-field-name='ProductName']");
        String productName = getString("productName");
        verifyTextElementUseTrim(productName , tableProduct);
    }
    /**
     * Siparişi onaya gönderme işlemi yapar. Çıkan pop-up onaylar
     */
    public void sendingForApprovalProcess() {
        page.locator("#SendApproveBtn").click();
        popUpConfirmationProcess();
    }
    /**
     * Siparişi onaylama işlemi yapar. Çıkan pop-up onaylar
     */
    public void approveOrder() {
        page.waitForLoadState(LoadState.DOMCONTENTLOADED);
        page.locator("#ApproveBtn").click();
        popUpConfirmationProcess();
        Assert.assertTrue(page.locator("#SetOrderGivenBtn").isEnabled());
    }

    /**
     * Siparişi verildi işlemi yapar. Çıkan pop-up onaylar
     */
    public void setOrderPlaced() {
        page.waitForLoadState(LoadState.DOMCONTENTLOADED);

        page.locator("#SetOrderGivenBtn").click();
        popUpConfirmationProcess();
    }
    /**
     * Siparişi sipariş no üzerinden doğrular
     */
    public void verifyOrderByOrderId() {
        page.waitForLoadState(LoadState.DOMCONTENTLOADED);

        String orderID = page.locator("#PurchaseOrderCode").getAttribute("value");
        Assert.assertNotNull("Order ID null olmamalı!", orderID);
    }

}

File: RetailTRUI.Tests/Pages/Purchasing/PurchaseOrderPage.cs
using RetailTRUI.Tests.Enums;
using RetailTRUI.Tests.Pages.Common;

namespace RetailTRUI.Tests.Pages.Purchasing;

/// <summary>
/// Purchase Order page object for creating and managing purchase orders
/// </summary>
public class PurchaseOrderPage : BasePage
{
    private ILocator PageTitle => Page.Locator("#PageTitle");
    private ILocator Calendar => Page.Locator(".k-icon.k-i-calendar");
    private ILocator SelectToday => Page.Locator(".k-link.k-nav-today");
    private ILocator ClickDropdownToggle => Page.Locator("span.k-select > span.k-icon.k-i-arrow-s");
    private IFrameLocator FrameLocator => Page.FrameLocator("iframe.k-content-frame");
    private IFrameLocator OrderProductIdentificationFrame => GetFrameByDialogTitle("Sipariş Ürünü Tanımlama");
    private IFrameLocator ProductDescriptionFrame => GetFrameByDialogTitle("Ürün Tanımlama");

    public async Task VerifyOrderCreationPageAsync()
    {
        var actualText = await PageTitle.TextContentAsync();
        actualText?.Trim().Should().Be("Sipariş İşlemleri");
    }

    public async Task FillOrderDateAsync()
    {
        await Calendar.Nth(0).ClickAsync();
        await SelectToday.ClickAsync(new LocatorClickOptions { Force = true });
    }

    public async Task FillOrderNameAsync()
    {
        var timestamp = DateTime.Now.ToString("yyyyMMddHHmm");
        var orderName = $"KMRN_TST_AUTO_{timestamp}";
        await Page.Locator("#PurchaseOrderName").FillAsync(orderName);
    }

    public async Task SelectCategoryFromListAsync(string category)
    {
        await OpenCategoryListAsync();
        await SelectCategoryAsync(category);
        await VerifyCategoryAsync(category);
    }

    private async Task OpenCategoryListAsync()
    {
        await ClickDropdownToggle.Nth(0).ClickAsync();
    }

    private async Task SelectCategoryAsync(string category)
    {
        var selectCategory = Page.Locator("#CategoryId_listbox li[role='option'].k-item")
            .Filter(new LocatorFilterOptions { HasText = category });
        await selectCategory.ClickAsync(new LocatorClickOptions { Force = true });
    }

    private async Task VerifyCategoryAsync(string category)
    {
        var selectedCategory = Page.Locator(".k-dropdown .k-input").Nth(0);
        var actualText = await selectedCategory.TextContentAsync();
        actualText?.Trim().Should().Be(category.Trim());
    }

    public async Task SetDistributorCompanyByCategoryAsync(string category)
    {
        await OpenCompanyIdentificationFrameAsync();
        await FillCompanyCodeAsync(category);
        await ClickFilterButtonIdAsync();
        await SelectDistributorCompanyAsync();
    }

    private async Task OpenCompanyIdentificationFrameAsync()
    {
        await Page.Locator("#FirmIdButtonId").ClickAsync();
    }

    private async Task FillCompanyCodeAsync(string category)
    {
        var categoryEnum = CategoriesExtensions.FromLabel(category);
        if (categoryEnum.HasValue)
        {
            var distributorInfo = categoryEnum.Value.GetDistributorInfo();
            var firmCode = distributorInfo.GetFirmCode();
            await FrameLocator.Locator("#FilterFirmCode").FillAsync(firmCode);
        }
    }

    private async Task ClickFilterButtonIdAsync()
    {
        await FrameLocator.Locator("#FilterButtonId").ClickAsync();
    }

    private async Task SelectDistributorCompanyAsync()
    {
        await FrameLocator.Locator("input[name^='FirmGridId']").Nth(0).ClickAsync();
    }

    public async Task SelectCompanyContactPersonAsync()
    {
        await ClickCompanyContactPersonAsync();
        await SelectContactPersonAsync();
    }

    private async Task ClickCompanyContactPersonAsync()
    {
        await Page.Locator(".k-multiselect-wrap.k-floatwrap").Nth(3).ClickAsync();
    }

    private async Task SelectContactPersonAsync()
    {
        await Page.WaitForSelectorAsync("#FirmResponsibleUserId_listbox");
        await Page.Locator("#FirmResponsibleUserId_option_selected").Nth(0).ClickAsync();
    }

    public async Task SelectDistributionTargetTypeAsync()
    {
        await OpenDistributionTargetTypeAsync();
        await SelectTargetTypeAsync();
    }

    private async Task OpenDistributionTargetTypeAsync()
    {
        await ClickDropdownToggle.Nth(1).ClickAsync(new LocatorClickOptions { Force = true });
    }

    private async Task SelectTargetTypeAsync()
    {
        await Page.WaitForSelectorAsync("#DistributionTargetTypeId_listbox li");
        await Page.Locator("#DistributionTargetTypeId_listbox li").Nth(1).ClickAsync();
    }

    public async Task SelectWarehouseWhereOrderWillEnterAsync(string region)
    {
        await OpenWarehouseDefinitionFrameAsync();
        await OpenSeturRegionFieldsAsync();
        await SelectSeturRegionFromListAsync(region);
        await VerifySeturRegionAsync(region);
        await ClickFilterButtonIdAsync();
        await SelectWarehouseAsync();
    }

    private async Task OpenWarehouseDefinitionFrameAsync()
    {
        await Page.Locator("#EntryWarehouseIdButtonId").ClickAsync();
    }

    private async Task OpenSeturRegionFieldsAsync()
    {
        await FrameLocator.Locator("span.k-select > span.k-icon.k-i-arrow-s").Nth(0).ClickAsync();
    }

    private async Task SelectSeturRegionFromListAsync(string region)
    {
        var allItems = FrameLocator.Locator("ul#FilterSeturRegionID_listbox li");
        var targetItem = allItems.Filter(new LocatorFilterOptions { HasText = region });
        await targetItem.ClickAsync(new LocatorClickOptions { Force = true });
    }

    private async Task VerifySeturRegionAsync(string region)
    {
        var selectedRegion = FrameLocator.Locator(".k-dropdown .k-input").Nth(0);
        var actualText = await selectedRegion.TextContentAsync();
        actualText?.Trim().Should().Be(region.Trim());
    }

    private async Task SelectWarehouseAsync()
    {
        await FrameLocator.Locator("input[name^='WarehouseGridId']").Nth(0).ClickAsync();
    }

    public async Task SelectInvoiceAddressAsync()
    {
        await OpenInvoiceAddressAsync();
        await SelectInvoiceAsync();
    }

    private async Task OpenInvoiceAddressAsync()
    {
        await ClickDropdownToggle.Nth(3).ClickAsync();
    }

    private async Task SelectInvoiceAsync()
    {
        await Page.WaitForSelectorAsync("#CompanyAddressId_listbox li");
        await Page.Locator("#CompanyAddressId_listbox li").Nth(1).ClickAsync();
    }

    public async Task SelectDeliveryAddressAsync()
    {
        await OpenDeliveryAddressAsync();
        await SelectDeliveryAsync();
    }

    private async Task OpenDeliveryAddressAsync()
    {
        await ClickDropdownToggle.Nth(4).ClickAsync();
    }

    private async Task SelectDeliveryAsync()
    {
        await Page.WaitForSelectorAsync("#WarehouseAddressId_listbox li");
        await Page.Locator("#WarehouseAddressId_listbox li").Nth(1).ClickAsync();
    }

    public async Task CompleteOrderAutomaticallyMarkCheckboxToNoAsync()
    {
        await Page.Locator("#no_CanAutoComplete").ClickAsync();
    }

    public async Task SaveOrderAsync()
    {
        await Page.Locator("#SaveBtn").ClickAsync();
        await VerifyPurchaseOrderTabsAsync();
        await VerifyOrderByOrderCodeAsync();
    }

    private async Task VerifyPurchaseOrderTabsAsync()
    {
        await Page.WaitForSelectorAsync("#PurchaseOrderTabs");
        var isVisible = await Page.Locator("#PurchaseOrderTabs").IsVisibleAsync();
        isVisible.Should().BeTrue();
    }

    private async Task VerifyOrderByOrderCodeAsync()
    {
        await Page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
        var orderID = await Page.Locator("#PurchaseOrderCode").GetAttributeAsync("value");
        orderID.Should().NotBeNullOrEmpty("Order ID should not be null!");
        AddString("orderCode", orderID!);
    }

    public async Task AddProductToOrderAsync()
    {
        await OpenOrderProductDescriptionFrameAsync();
        await OpenProductDescriptionFrameAsync();
        await GetProductNameAsync();
        await SelectProductAsync();
        await GetCurrencyCodesAsync();

        if (await IfCurrencyNotMatchCloseFrameAsync())
        {
            await OpenOrderCurrencyCodesAsync();
            await SelectCurrencyCodeAsync();
            await VerifyProductCurrencyCodeAsync();
            await SaveOrderAsync();
            await ConfirmPopupAsync();

            await OpenOrderProductDescriptionFrameAsync();
            await OpenProductDescriptionFrameAsync();
            await SelectProductAsync();
        }

        await EnterQuantityForProductAsync();
        await SaveOrderProductsDescriptionAsync();
    }

    private async Task OpenOrderProductDescriptionFrameAsync()
    {
        await Page.Locator("a.k-grid-PurchaseOrderProductGridIdAddNew").ClickAsync();
    }

    private async Task OpenProductDescriptionFrameAsync()
    {
        await FrameLocator.Locator("#ProductIdButtonId").ClickAsync();
    }

    private async Task GetProductNameAsync()
    {
        var productNameLocator = ProductDescriptionFrame.Locator("td[data-field-name='ProductNameLong']").Nth(0);
        await productNameLocator.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Visible });
        var productName = (await productNameLocator.TextContentAsync())?.Trim();
        AddString("productName", productName!);
    }

    private async Task SelectProductAsync()
    {
        await ProductDescriptionFrame.Locator("input[name^='ProductGrid']").Nth(0).ClickAsync();
    }

    private async Task GetCurrencyCodesAsync()
    {
        var currencyLocator = OrderProductIdentificationFrame.Locator("#ProductCurrencyCode");
        await currencyLocator.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Visible });

        var productCurrency = await currencyLocator.EvaluateAsync<string>(
            "el => el.value || el.placeholder || el.textContent || el.innerText");

        AddString("productCurrency", productCurrency);
    }

    private async Task<bool> IfCurrencyNotMatchCloseFrameAsync()
    {
        var currencyCode = await OrderProductIdentificationFrame.Locator("#CurrencyCode").InputValueAsync();
        var productCurrency = GetString("productCurrency");

        if (currencyCode != productCurrency)
        {
            await Page.Locator(".k-window-actions .k-window-action").Nth(0).ClickAsync();
            return true;
        }
        return false;
    }

    private async Task OpenOrderCurrencyCodesAsync()
    {
        var currencyDropdownToggle = Page.Locator("span.k-select > span.k-icon.k-i-arrow-s");
        await currencyDropdownToggle.Nth(7).HoverAsync();
        await currencyDropdownToggle.Nth(7).ClickAsync();
    }

    private async Task SelectCurrencyCodeAsync()
    {
        var productCurrency = GetString("productCurrency");
        var currencyOption = Page.Locator("ul#CurrencyCode_listbox li")
            .Filter(new LocatorFilterOptions { HasText = productCurrency });
        await currencyOption.ClickAsync(new LocatorClickOptions { Timeout = 3000, Force = true });
    }

    private async Task VerifyProductCurrencyCodeAsync()
    {
        var selectedCurrency = Page.Locator(".k-dropdown .k-input").Nth(3);
        var productCurrency = GetString("productCurrency");
        var actualText = await selectedCurrency.TextContentAsync();
        actualText?.Trim().Should().Be(productCurrency.Trim());
    }

    private async Task EnterQuantityForProductAsync()
    {
        var randomQuantity = Random.Shared.Next(1, 21);
        var quantityLocator = OrderProductIdentificationFrame.Locator("#Quantity");
        
        await quantityLocator.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Attached });
        var handle = await quantityLocator.ElementHandleAsync();
        
        if (handle != null)
        {
            await Page.EvaluateAsync(@"
                (element, val) => {
                    const widget = $(element).data('kendoNumericTextBox');
                    if (widget) {
                        widget.value(val);
                        widget.trigger('change');
                    }
                }", new object[] { handle, randomQuantity });
        }

        await OrderProductIdentificationFrame.Locator("#Quantity").PressAsync("Enter");
    }

    private async Task SaveOrderProductsDescriptionAsync()
    {
        await OrderProductIdentificationFrame.Locator("#SaveBtn").ClickAsync();
    }

    public async Task VerifyProductsAddedToOrderAsync()
    {
        var tableProduct = Page.Locator("td[data-field-name='ProductName']");
        var productName = GetString("productName");
        var actualText = await tableProduct.TextContentAsync();
        actualText?.Trim().Should().Be(productName.Trim());
    }

    public async Task SendingForApprovalProcessAsync()
    {
        await Page.Locator("#SendApproveBtn").ClickAsync();
        await PopUpConfirmationProcessAsync();
    }

    public async Task ApproveOrderProcessAsync()
    {
        await Page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
        await Page.Locator("#ApproveBtn").ClickAsync();
        await PopUpConfirmationProcessAsync();
        
        var isEnabled = await Page.Locator("#SetOrderGivenBtn").IsEnabledAsync();
        isEnabled.Should().BeTrue();
    }

    public async Task SetOrderPlacedAsync()
    {
        await Page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
        await Page.Locator("#SetOrderGivenBtn").ClickAsync();
        await PopUpConfirmationProcessAsync();
    }

    public async Task VerifyOrderByOrderIdAsync()
    {
        await Page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
        var orderID = await Page.Locator("#PurchaseOrderCode").GetAttributeAsync("value");
        orderID.Should().NotBeNullOrEmpty("Order ID should not be null!");
    }

    /// <summary>
    /// Helper method to complete full order process for a category
    /// </summary>
    public async Task OrderCompletionAsync(string category)
    {
        var globalPage = new GlobalPage();
        await globalPage.ClickPurchaseOrderCreationLinkAsync();
        
        await VerifyOrderCreationPageAsync();
        await FillOrderDateAsync();
        await FillOrderNameAsync();
        await SelectCategoryFromListAsync(category);
        await SetDistributorCompanyByCategoryAsync(category);
        await SelectCompanyContactPersonAsync();
        await SelectDistributionTargetTypeAsync();
        await SelectWarehouseWhereOrderWillEnterAsync("KAPIKULE-SANAL");
        await SelectInvoiceAddressAsync();
        await SelectDeliveryAddressAsync();
        await CompleteOrderAutomaticallyMarkCheckboxToNoAsync();
        await SaveOrderAsync();
        await AddProductToOrderAsync();
        await VerifyProductsAddedToOrderAsync();
        await SendingForApprovalProcessAsync();
        await ApproveOrderProcessAsync();
        await SetOrderPlacedAsync();
        await VerifyOrderByOrderIdAsync();
    }

    /// <summary>
    /// Helper method to set proforma and invoice for order
    /// Used in invoice completion tests
    /// </summary>
    public async Task SetProformaAndInvoiceAsync()
    {
        var globalPage = new GlobalPage();
        var orderSearchPage = new PurchaseOrderSearchPage();
        
        await globalPage.ClickPurchasingDropdownToggleAsync();
        await globalPage.ClickPurchaseOrderSearchLinkAsync();
        await orderSearchPage.VerifyPurchaseOrderSearchPageAsync();
        await orderSearchPage.SearchOrderByOrderNumberAndEditOrderAsync();
        await orderSearchPage.GoToOrderProformasTabAsync();
        await orderSearchPage.AddInfoForProformaAndSaveAsync();
        await orderSearchPage.CopyOrderItemsAndApproveProformaAsync();
        await orderSearchPage.GoToOrderInvoicesTabAsync();
        await orderSearchPage.AddInfoForInvoiceAndSaveAsync();
        await orderSearchPage.CopyProformaItemsAndApproveInvoiceAsync();
        await orderSearchPage.CompletingAndApprovingInvoiceAsync();
    }
}
